types.singleLineStr: strings that don't contain '\n'

Julow opened the pull request:
###### Motivation for this change

Add a new type, inheriting `types.str`, but checking whether the value doesn't contain any newline characters.

The motivation comes from a problem with the `users.users.${u}.openssh.authorizedKeys` option.
It is easy to unintentionally insert a newline character at the end of a string, or even in the middle, for example:

```nix
restricted_ssh_keys = command: keys:
  let
    prefix = ''
      command="${command}",no-pty,no-agent-forwarding,no-port-forwarding,no-X11-forwarding
    '';
  in map (key: "${prefix} ${key}") keys;
```

The `prefix` string ends with a newline, which ends up in the middle of a key entry after a few manipulations.

This is problematic because the key file is built by concatenating all the keys with `concatStringsSep "\n"`, with result in two entries for the faulty key:

```nix
''
  command="...",options...
  MY_KEY
''
```

This is hard to debug and might be dangerous. This is now caught at build time.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Julow committed:
types.singleLineStr: strings that don't contain '\n'

Add a new type, inheriting 'types.str' but checking whether the value
doesn't contain any newline characters.

The motivation comes from a problem with the
'users.users.${u}.openssh.authorizedKeys' option.
It is easy to unintentionally insert a newline character at the end of a
string, or even in the middle, for example:

    restricted_ssh_keys = command: keys:
      let
        prefix = ''
          command="${command}",no-pty,no-agent-forwarding,no-port-forwarding,no-X11-forwarding
        '';
      in map (key: "${prefix} ${key}") keys;

The 'prefix' string ends with a newline, which ends up in the middle of
a key entry after a few manipulations.

This is problematic because the key file is built by concatenating all
the keys with 'concatStringsSep "\n"', with result in two entries for
the faulty key:

    ''
      command="...",options...
      MY_KEY
    ''

This is hard to debug and might be dangerous. This is now caught at
build time.

pennae reviewed (commented):
this looks better than the first attempt! :) much better chance of getting a helpful error message out of it.

not sure the extra type (in this form) is needed though. `strMatching "[^\n]*"` should work just as well (and be faster during eval). maybe the extra type is warranted if there's enough other instances in tree, haven't checked that yet.

roberth commented:
I think we should also disallow `\r` for good measure.

> not sure the extra type (in this form) is needed though.

I find the `description` more natural.

>  strMatching "[^\n]*" should work just as well (and be faster during eval).

Then the type could be based off that. I think the file has some examples of using `//` for this purpose.

pennae commented:
> I find the description more natural.

it is! not arguing against the description, just against including it in `types.nix` if it's not used anywhere else. but it *does* feel like something that could be useful elsewhere, so eh let's add it :)

Julow committed:
types.singleLineStr: Disallow \r

Julow commented:
Just added the `\r` and re-implemented it on top of `singleLineStr` but keeping it defined in `types`. I'm in favor of keeping it defined because I feel it might be used in other places (everywhere that takes a `listOf str` and does `concatStringsSep "\n"`).

pennae commented:
just remembered another case where this might be useful: wireguard key attrs #119312. so this definitely is a good thing to have :)

Julow commented:
I can add a commit for the wireguard issue.

pennae commented:
oh, that wasn't supposed to be a request! :sweat_smile: just me contradicting my own point that there's only one place where it'd be useful. but if you want to fix wireguard as well, please do! :D

Julow commented:
I have a commit [here](https://github.com/Julow/nixpkgs/tree/single_line_str_wireguard) for the specific option but I notice that many options in wireguard are supposed to be single line.

I guess it's easier to impose an option type everywhere than it is to ban string concatenations. But perhaps the wireguard case should be discussed a bit more. Other solutions might be to trim the end of strings more often or always or to have a `concatLines` function that asserts that every lines don't contain any `\n`.

roberth commented:
> trim the end of strings

You could do so in the type by adding a conversion after the merge function; see `coerceTo`'s merge for an example.
It makes sense to do so, because of the prevalent "end line" interpretation of `\n`, as opposed to "new line".

pennae commented:
:+1: for automatically trimming a trailing `\n` in `singleLineStr` *if* it's the only `\n`. that would neatly fix the wireguard example, and `''`-quoted strings as well.

Julow commented:
That's a good idea ! I see no downside and it makes the type work better with `readFile` and `''`-strings. I'll work on that.

Julow committed:
types.singleLineStr: Allow and trim trailing \n

Allow a \n character at the end of the string and remove it during the
merge function.

An option of this type will resolve to the value "foo" whether it is set
to "foo" or "foo\n".

This is useful when using 'builtins.readFile' or ''-strings, which might
add an unintended newline (for example, bash trim the final newline from
a subshell).

Julow commented:
I've changed the type to allow a trailing `\n` and to trim it during the merge function.

That the merge function is called even when the option is set once is counter-intuitive but after some reading it seems to be true and used a lot by other types.

pennae commented on lib/types.nix:
```diff
@@ -300,6 +300,19 @@ rec {
+      in
+      mkOptionType {
+        name = "singleLineStr";
+        description = "string that doesn't contain [\\n\\r]";
```
this isn't quite correct anymore and could be misleading, but not sure how to phrase it well without introducing odd artifacts in composite types (eg how `listOf` uses a `list of ${elem}s` as its description)

maybe something like this?
```suggestion
        description = "(optionally [\\r\\n]-terminated) single-line string";
```

Julow replied:
Thanks :)

pennae replied:
oh heck, our suggestion was wrong. you're only checking for a `\n` terminator, not for `[\r\n]`. didn't even notice that until now, when we wrote it we thought you checked them both. :(

Julow replied:
I wondered about matching both but I saw that everywhere in nixpkgs there were splits and concats with just `\n` and very few occurrences of `\r` so I thought not allowing `\r` inside the string is relatively harmless (and there's a few occurrences of that in nixpkgs) but actually expecting a `\r\n` at the end would be a first.

pennae replied:
yeah, expecting an `\n` at the end is totally fine. should've just written `newline-terminated` instead of `[\r\n]-terminated` :D

Julow replied:
I amended that.

Julow committed:
types.singleLineStr: Improve description

Co-authored-by: pennae <82953136+pennae@users.noreply.github.com>

roberth commented:
Great, thanks @Julow!
